content/_code-samples/tx-serialization/js/index.js

```diff
@@ -0,0 +1,687 @@
+        const ISO_REGEX = /^[A-Z0-9a-z?!@#$%^&*(){}[\]|]{3}$/
+        const HEX_REGEX = /^[A-F0-9]{40}$/
+
+        if(ISO_REGEX.test(codeString)) {
+            if(codeString === "XRP") {
+                if (isXrpOk) {
+                    // Rare, but when the currency code "XRP" is serialized, it's
+                    // a special-case all zeroes.
+                    logger("Currency code(XRP): " + "00".repeat(20))
+                    return "00".repeat(20)
+                }
+
+                throw new Error("issued currency can't be XRP")
+            }
+            const codeAscii = Buffer.from(codeString, 'ascii')
+            logger("Currency code ASCII: " + codeAscii.toString("hex"))
+            // standard currency codes: https://xrpl.org/currency-formats.html#standard-currency-codes
+            // 8 bits type code (0x00)
+            // 88 bits reserved (0's)
+            // 24 bits ASCII
+            // 16 bits version (0x00)
+            // 24 bits reserved (0's)
+            const prefix = Buffer.alloc(12)
+            const postfix = Buffer.alloc(5)
+
+            return Buffer.concat([prefix, codeAscii, postfix]).toString("hex")
+        } else if (HEX_REGEX.test(codeString)) {
+            // raw hex code
+            return Buffer.from(codeString).toString("hex")
+        }
+
+        throw new Error("invalid currency code")
+    }
+
+    /**
+     *    Serializes a hexadecimal string as binary and confirms that it's 128 bits
+     *
+     * @param contents
+     * @returns {string}
+     */
+    hash128ToBytes(contents) {
+        const buffer = this.hashToBytes(contents)
+        if(buffer.length !== 16) {
+            // 16 bytes = 128 bits
+            throw new Error("Hash128 is not 128 bits long")
+        }
+
+        return buffer.toString("hex")
+    }
+
+    /**
+     *    Serializes a hexadecimal string as binary and confirms that it's 160 bits
+     *
+     * @param contents
+     * @returns {string}
+     */
+    hash160ToBytes(contents) {
+        const buffer = this.hashToBytes(contents)
+        if(buffer.length !== 20) {
+            // 20 bytes = 160 bits
+            throw new Error("Hash160 is not 160 bits long")
+        }
+
+        return buffer.toString("hex")
+    }
+
+    /**
+     *    Serializes a hexadecimal string as binary and confirms that it's 128 bits
+     *
+     * @param contents
+     * @returns {string}
+     */
+    hash256ToBytes(contents) {
+        const buffer = this.hashToBytes(contents)
+        if(buffer.length !== 32) {
+            // 32 bytes = 256 bits
+            throw new Error("Hash256 is not 256 bits long")
+        }
+
+        return buffer.toString("hex")
+    }
+
+    /**
+     *     Helper function; serializes a hash value from a hexadecimal string
+     *     of any length.
+     *
+     * @param contents
+     * @returns {string}
+     */
+    hashToBytes(contents) {
+        return Buffer.from(contents).toString("hex")
```

Comment:
Not a big deal but there is an edge case where if the contents is empty it should be zero bytes.

See https://github.com/XRPLF/xrpl.js/pull/2156/files

Reply:
@ckniffen Hi Caleb, just for clarification: The expected behavior is that an empty hash should NOT return a byte string of zeros the appropriate length but an empty string so that a field can be unset - is this correct? "Zero bytes" is a homonym here :)

Reply:
So, if I get this right, an empty input should return an zero-byte padded string of hash-type-length, correct?

Reply:
Yeah.  The only case this needs to happen is for clearing an Email Hash.  So not the most important thing to demo

Reply:
> Two more recommendations:
> 
> 1. I'd recommend capitalizing the serialized sub-components as you log them to match the eventual output:
>    Ex.
>    `Fee: 400000000000000a` -> `Fee: 400000000000000A`
> 2. I'd recommend printing out the transaction you're serializing at the top, then show the serialized one below, with a couple lines explaining what's happening. (This will affect your tests)
> 
> The reason I'd recommend that is currently if someone just does `node index.js` it's quite hard to tell what's happening, because it just shows the serialized output without any context.
> 
> But if you add a couple print statements for context they can figure out what happened from just the console logs.
> 
> Ex.
> 
> ```
> Serializing this transaction: 
> {
>   "Account": "rMBzp8CgpE441cp5PVyA9rpVV7oT8hP3ys",
>   "Expiration": 595640108,
>   "Fee": "10",
>   "Flags": 524288,
>   "OfferSequence": 1752791,
>   "Sequence": 1752792,
>   "SigningPubKey": "03EE83BB432547885C219634A1BC407A9DB0474145D69737D09CCDC63E1DEE7FE3",
>   "TakerGets": "15000000000",
>   "TakerPays": {
>     "currency": "USD",
>     "issuer": "rvYAfWj5gh67oV6fW32ZzP3Aw4Eubs59B",
>     "value": "7072.8"
>   },
>   "TransactionType": "OfferCreate",
>   "TxnSignature": "30440220143759437C04F7B61F012563AFE90D8DAFC46E86035E1D965A9CED282C97D4CE02204CFD241E86F17E011298FC1A39B63386C74306A5DE047E213B0F29EFA4571C2C",
>   "hash": "73734B611DDA23D3F5F62E20A173B78AB8406AC5015094DA53F53D39B9EDB06C"
> }
> 
> Serialized Transaction: 120007220008000024001ABED82A2380BF2C2019001ABED764D55920AC9391400000000000000000000000000055534400000000000A20B3C85F482532A9578DBB3950B85CA06594D165400000037E11D60068400000000000000A732103EE83BB432547885C219634A1BC407A9DB0474145D69737D09CCDC63E1DEE7FE3744630440220143759437C04F7B61F012563AFE90D8DAFC46E86035E1D965A9CED282C97D4CE02204CFD241E86F17E011298FC1A39B63386C74306A5DE047E213B0F29EFA4571C2C8114DD76483FACDEE26E60D8A586BB58D09F27045C46
> ```

No problem, I just stuck to the py version output 1:1. So I will improve that one so it looks good and reads well, and if it gets approved perhaps I should do a quick update/PR for the py version also to have it reflect it.